src/mono/wasm/runtime/hybrid-globalization/normalization.ts

```diff
@@ -0,0 +1,73 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { setU16 } from "../memory";
+import { mono_wasm_new_external_root } from "../roots";
+import { conv_string_root } from "../strings";
+import { MonoString, MonoStringRef } from "../types";
+import { Int32Ptr } from "../types/emscripten";
+import { pass_exception_details } from "./common";
+
+export function mono_wasm_is_normalized(exceptionMessage: Int32Ptr, normalizationForm: number, inputStr: MonoStringRef) : number{
+    const inputRoot = mono_wasm_new_external_root<MonoString>(inputStr);
+    try{
+        const jsString = conv_string_root(inputRoot);
+        if (!jsString)
+            throw new Error("Invalid string was received.");
+
+        const normalization = normalization_to_string(normalizationForm);
+        const result = jsString.normalize(normalization);
+        return result === jsString ? 1 : 0;
+    }
+    catch (ex) {
+        pass_exception_details(ex, exceptionMessage);
+        return -1;
+    } finally {
+        inputRoot.release();
+    }
+}
+
+export function mono_wasm_normalize_string(exceptionMessage: Int32Ptr, normalizationForm: number, inputStr: MonoStringRef, dstPtr: number, dstLength: number) : number{
+    const inputRoot = mono_wasm_new_external_root<MonoString>(inputStr);
+    try {
+        const jsString = conv_string_root(inputRoot);
+        if (!jsString)
+            throw new Error("Invalid string was received.");
+
+        const normalization = normalization_to_string(normalizationForm);
+        const result = jsString.normalize(normalization);
+
+        // increase the dest buffer
+        if (result.length > dstLength)
+            return result.length;
+        for (let i = 0; i < result.length; i++)
+            setU16(dstPtr + i*2, result.charCodeAt(i));
```

Comment:
why this is not TextEncoder ?

Reply:
In https://github.com/dotnet/runtime/pull/85516 I tested how `TextEncoder` works and it turned to be slower than this loop.

Reply:
Maybe creating the `TextEncoder` instance is slow, but the actual operation fast?

Reply:
Encoding's logic is similar to the decoding logic, so we are creating `StringEncoder` only once and then calling `TextEncoder.encode`. To see how the timing behaves with string size change I created the comparison. 

With TextEncoder:

| measurement | time 64kB | time 128kB | time 256kB | time 512kB |
|-:|-:|-:|-:|-:|
|               String, TextInfo ToLower |     2.3724ms |     4.9690ms |     8.7664ms |    19.8859ms |
|               String, TextInfo ToUpper |     2.6132ms |     4.9603ms |     8.5550ms |    20.2031ms |
|            String, TextInfo ToTileCase |    51.4961ms |    89.3059ms |   179.8772ms |   381.3721ms |

Without TextEncoder:

| measurement | time 64kB | time 128kB | time 256kB | time 512kB |
|-:|-:|-:|-:|-:|
|               String, TextInfo ToLower |     0.8991ms |     1.8787ms |     4.5523ms |     8.1296ms |
|               String, TextInfo ToUpper |     1.0265ms |     2.2138ms |     4.4705ms |     8.0644ms |
|            String, TextInfo ToTileCase |    46.9924ms |   103.1250ms |   200.0556ms |   386.0698ms |

Both are linear and we cannot see much TextEncoder instance creation impact, but the latter option is visibly better.

Reply:
Maybe the `TextEncoder` does more work ? Like processing multi word characters correctly ?

Reply:
In our tests with multiple codepoints they work the same with the exception that TextEncoder/Decoder cannot handle "invalid char" symbols:
- [U+D83C](https://www.fileformat.info/info/unicode/char/d83c/index.htm)
- [U+DF08](https://www.fileformat.info/info/unicode/char/DF08/index.htm)
they change both to [U+FFFD](https://www.fileformat.info/info/unicode/char/FFFD/index.htm).